### PR TITLE
Speech Characteristics v3.0

### PR DESCRIPTION
**Pull request by GeorgeEfstathiadis**
Methods doc: https://docs.google.com/document/d/1d3-lZyN7HO3ROa6sBewEo4yqzMVTyCngRBzWEQl_2Iw/edit

### CODE REVIEW

**Review by vjbytes102 — APPROVED** [review]

File: openwillis/measures/text/speech_attribute.py [hunk 1]
```diff
@@ -374,8 +355,11 @@ def speech_characteristics(json_conf, language="en", speaker_label=None, min_tur
 
         if bool(json_conf):
             language = language[:2].lower() if (language and len(language) >= 2) else "na"
+            if language not in measures["english_langs"] + measures["supported_langs_sentence_embeddings"] + measures["supported_langs_bert"]:
+                logger.warning(f"Language {language} not supported. Defaulting to English")
+                language = "en"
```
Comment (vjbytes102) [inline openwillis/measures/text/speech_attribute.py:360]: Just to confirm: We will compute all the language-dependent variables if a user passes an unsupported language code, regardless of the original language

File: openwillis/measures/text/util/characteristics_util.py [hunk 2]
```diff
@@ -679,63 +732,212 @@ def get_mattr(text):
...
+
+    This function calculates measures related to the first person pronouns in each turn.
+
+    Parameters:
+    ...........
+    turn_df: pandas dataframe
+        A dataframe containing turn summary information.
+    turn_list: list
+        List of transcribed text at the turn level.
+
```
Comment (vjbytes102) [inline openwillis/measures/text/util/characteristics_util.py:858]: add description for measures

File: openwillis/measures/text/util/characteristics_util.py [hunk 3]
```diff
@@ -800,6 +1008,492 @@ def get_sentiment(df_list, text_list, measures):
...
+        word_embeddings = np.concatenate(word_embeddings, axis=0)
+    else:
+
+        inputs = tokenizer(word_list, return_tensors='pt', padding=True)
+        outputs = model(**inputs)
+        # Average pooling of the hidden states
+        word_embeddings = outputs.last_hidden_state.mean(1).detach().numpy()
+    return word_embeddings
+
+def get_word_coherence(df_list, utterances_speaker, language, measures):
```
Comment (vjbytes102) [inline openwillis/measures/text/util/characteristics_util.py:1143]: Not urgent: this function is huge, can we break it for code scalability

File: openwillis/measures/text/config/text.json [hunk 4]
```diff
@@ -47,5 +47,72 @@
...
+  "turn_to_turn_tangeniality_slope": "turn_to_turn_tangeniality_slope",
+  "panss_string": "[Hi I am We are going to be spending the next 30 to 40 minutes talking about you and your reasons for being here. Maybe you can start out by telling me something about yourself and your background,Have you been feeling worried or nervous in the past week,Would you say that you are usually calm and relaxed,What is been making you feel nervous,Just how nervous have you been feeling,Have you been shaking at times or has your heart been racing,Do you get into a state of panic,Has your sleep eating or participation in activities been affected,Have things been going well for you,Has anything been bothering you lately,Can you tell me something about your thoughts on life and its purpose,Do you follow a particular philosophy any special rules teachings or religious doctrine,Some people tell me they believe in the Devil what do you think,Can you tell me more about this,Can you read other people minds, How does that work,Can others read your mind,How can they do that,Is there any reason that someone would want to read your mind,Who controls your thoughts]",
```
[the remark this answers is not in the document]
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/config/text.json:114]: actually I don't think we're even using it in speech characteristics, so I will just remove it from the text.json

File: openwillis/measures/text/speech_attribute.py (hunk 1 above)
Comment (GeorgeEfstathiadis) [reply openwillis/measures/text/speech_attribute.py:360]: that's what I thought we could do yes. If they input a supported language that is not english, we compute only the non-english needed measures, if they input an unsupported language code we act as if the language is English. But maybe it makes more sense to throw an error and stop the calculation. I'm not sure what do you think?

**Review by vjbytes102 — APPROVED** [review]

**Pull request merged**